src/cli/filter_arg.zig

```diff
@@ -0,0 +1,302 @@
+            .e_array => |arr| arr,
+            .e_object => |obj| if (obj.get("packages")) |packages| switch (packages.data) {
+                .e_array => |arr| arr,
+                else => break,
+            } else break,
+            else => break,
+        };
+
+        for (json_array.slice()) |expr| {
+            switch (expr.data) {
+                .e_string => |pattern_expr| {
+                    // /basepath/pattern/package.json
+                    // const size = parent_trimmed.len + 1 + pattern_expr.data.len + "/package.json".len;
+                    const size = pattern_expr.data.len + "/package.json".len;
+                    var pattern = try allocator.alloc(u8, size);
+                    @memcpy(pattern[0..pattern_expr.data.len], pattern_expr.data);
+                    @memcpy(pattern[pattern_expr.data.len..size], "/package.json");
+
+                    // @memcpy(pattern[0..parent_trimmed.len], parent_trimmed);
+                    // pattern[parent_trimmed.len] = '/';
+                    // @memcpy(pattern[parent_trimmed.len + 1 .. parent_trimmed.len + 1 + pattern_expr.data.len], pattern_expr.data);
+                    // @memcpy(pattern[parent_trimmed.len + 1 + pattern_expr.data.len .. size], "/package.json");
+                    try out_patterns.append(pattern);
+                },
+                else => {
+                    // TODO log error and fail
+                    Global.crash();
+                },
+            }
+        }
+        @memcpy(root_buf[0..parent_trimmed.len], parent_trimmed);
+        return root_buf[0..parent_trimmed.len];
+    }
+
+    // if we were not able to find a workspace root, we simply glob for all package.json files
+    try out_patterns.append(try allocator.dupe(u8, "**/package.json"));
+    const root_dir = strings.withoutTrailingSlash(workdir_);
+    @memcpy(root_buf[0..root_dir.len], root_dir);
+    return root_buf[0..root_dir.len];
+}
+
+pub const FilterSet = struct {
+    allocator: std.mem.Allocator,
+    filters: []Pattern,
+    has_name_filters: bool = false,
+
+    const Pattern = struct {
+        codepoints: []u32,
+        kind: enum {
+            name,
+            path,
+        },
+        // negate: bool = false,
+    };
+
+    pub fn init(allocator: std.mem.Allocator, filters: []const []const u8, cwd: []const u8) !FilterSet {
+        var buf: bun.PathBuffer = undefined;
+        // TODO fixed buffer allocator with fallback?
+        var self = FilterSet{ .allocator = allocator, .filters = try allocator.alloc(Pattern, filters.len) };
+        for (0.., filters) |idx, filter_utf8_| {
+            var filter_utf8 = filter_utf8_;
+            const is_path = filter_utf8.len > 0 and filter_utf8[0] == '.';
+            if (is_path) {
+                const parts = [_]string{filter_utf8};
+                filter_utf8 = bun.path.joinAbsStringBuf(cwd, &buf, &parts, .auto);
+            }
+            var filter_utf32 = std.ArrayListUnmanaged(u32){};
```

Comment:
feels like this will be very slow but i guess the filters won't be long so it should be fine